File: leetcode-cn/sword2offer/000剑指0_Offer_42._连续子数组的最大和.py
'''
执行用时：56 ms, 在所有 Python3 提交中击败了94.57% 的用户
内存消耗：18.5 MB, 在所有 Python3 提交中击败了71.64% 的用户
'''
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maximum = nums[0]
        pre = 0
        for num in nums:
            pre = max(pre + num, num)
            maximum = max(maximum, pre)
        return maximum


'''
执行用时：60 ms, 在所有 Python3 提交中击败了90.07% 的用户
内存消耗：19.2 MB, 在所有 Python3 提交中击败了48.62% 的用户
'''
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        for i in range(1, len(nums)):
            nums[i] += max(nums[i - 1], 0)
        return max(nums)

File: leetcode-cn/sword2offer/000剑指0_Offer_42._连续子数组的最大和.cpp
/*
DP
Time: O(N)
Space: O(1)

执行用时：24 ms, 在所有 C++ 提交中击败了66.42% 的用户
内存消耗：22.4 MB, 在所有 C++ 提交中击败了46.23% 的用户
*/
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maximum = nums[0], pre = 0;
        for (auto num: nums) {
            pre = max(pre + num, num);
            maximum = max(maximum, pre);
        }
        return maximum;
    }
};
